### PR TITLE
feat: security-policy for ic-assets.json5

### PR DESCRIPTION
**Pull request by sesi200**
# Description

Introduces the `"security_policy"` and `"disable_security_policy_warning"` fields to `.ic-assets.json5`.

Fixes [SDK-1718](https://dfinity.atlassian.net/browse/SDK-1718)

# How Has This Been Tested?

Added e2e

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

[SDK-1718]: https://dfinity.atlassian.net/browse/SDK-1718?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,31 @@
...
+If `"standard"` is specified, a set of security headers is added to the asset. The headers can be displayed with `dfx info security-policy`. It is a set of security headers that will work for most dapps. A warning is displayed that the headers could be hardened.
+
+If `"hardened"` is set, the same headers as with `"standard"` are added. However, the asset sync expects that improved headers are set that would improve security where appropriate and therefore does not display the warning.
+
+All warnings regarding security policies can be disabled with ``"disable_security_policy_warning": true`. It needs to be set per asset.
+
+The standard security policy headers can be seen with `dfx info security-policy`.
+It also contains a lot of suggestions on how to harden the policy.
+
+Updated the starter projects to use `"security_policy"` instead of including the whole security policy.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:23]:
```suggestion
Updated the starter projects to use `"security_policy"` instead of including the whole security policy by defining individual headers.
```

File: docs/cli-reference/dfx-info.mdx [hunk 2]
```diff
@@ -14,13 +14,14 @@ dfx info [type] [flag]
...
-| replica-port        | The listening port of the replica.             |
-| replica-rev         | The revision of the bundled replica.           |
-| webserver-port      | The local webserver (icx-proxy) port.          |
+| Information         | Description                                                                                                   |
+|---------------------|---------------------------------------------------------------------------------------------------------------|
+| candid-ui-url       | The URL of the Candid UI canister.                                                                            |
+| networks-json-path  | Path to network definition file networks.json.                                                                |
+| replica-port        | The listening port of the replica.                                                                            |
+| replica-rev         | The revision of the bundled replica.                                                                          |
+| security-policy     | Show the standard headers that gets applied to assets in .ic-assets.json5 if "security_policy" is "standard". |
```
Comment (ericswanson-dfinity) [inline docs/cli-reference/dfx-info.mdx:23]: These are also the headers defined if security policy is `hardened`, right?

File: e2e/tests-dfx/info.bash [hunk 3]
```diff
@@ -66,3 +66,25 @@ teardown() {
...
+  touch src/e2e_project_frontend/assets/thing.json
+
+  dfx_start
+  dfx canister create --all
+
+  echo "[
+    {
+      \"match\": \"**/*\",
+      \"headers\": {
+        $(dfx info security-policy)
```
Comment (ericswanson-dfinity) [inline e2e/tests-dfx/info.bash:82]: Great test!  However, this test would pass if `dfx info security-policy` produced no output to stdout.  Could we have a separate test that digs into its output a little bit?

File: src/canisters/frontend/ic-asset/src/security_policy.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+use std::fmt::Display;
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+use crate::asset::config::HeadersConfig;
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Copy, Debug)]
+#[serde(rename_all = "lowercase")]
+/// Asset synchronization will warn if no
```
Comment (ericswanson-dfinity) [inline src/canisters/frontend/ic-asset/src/security_policy.rs:12]: "will warn if no" ...

is this sentence incomplete?

File: src/canisters/frontend/ic-asset/src/sync.rs [hunk 5]
```diff
@@ -306,6 +308,64 @@ pub(crate) fn gather_asset_descriptors(
...
+            } else {
+                warn!(logger, "Unhardened assets:");
+                for asset in &standard_policy_assets {
+                    warn!(logger, "  - {}", asset.key);
+                }
+            }
+        }
+        if !standard_policy_assets.is_empty() || !no_policy_assets.is_empty() {
+            warn!(logger, "To disable the policy warning, define \"disable_standard_security_policy_warning\": true in .ic-assets.json5.");
+        }
```
Comment (ericswanson-dfinity) [inline src/canisters/frontend/ic-asset/src/sync.rs:368]: The only thing I see missing here is a warning for assets that are marked as hardened, but that aren't actually hardened (don't define any headers, whether new or updates to existing).

It seems to me that it would be natural to think that by changing the security policy from `standard` to `hardened` would mean using a more hardened set of security headers, even though there is documentation to the contrary.  But `hardened` actually means "continue to use the standard security policy headers and don't warn me about it"

In other words, suggested avenues to quiet the warnings:
- `standard` or `disabled` security policy and `disable_security_policy_warning` set to true
- `hardened` security policy and `headers` has at least one element

I would further suggest that `hardened` security policy without at least one `headers` element (ostensibly "hardening" it) should be an error, not a warning

File: src/canisters/frontend/ic-asset/src/sync.rs [hunk 6]
```diff
@@ -306,6 +308,64 @@ pub(crate) fn gather_asset_descriptors(
...
+                warn!(logger, "Unhardened assets: all");
+            } else {
+                warn!(logger, "Unhardened assets:");
+                for asset in &standard_policy_assets {
+                    warn!(logger, "  - {}", asset.key);
+                }
+            }
+        }
+        if !standard_policy_assets.is_empty() || !no_policy_assets.is_empty() {
+            warn!(logger, "To disable the policy warning, define \"disable_standard_security_policy_warning\": true in .ic-assets.json5.");
```
Comment (ericswanson-dfinity) [inline src/canisters/frontend/ic-asset/src/sync.rs:367]: If I'm reading the config.rs right...
```suggestion
            warn!(logger, "To disable the policy warning, define \"disable_security_policy_warning\": true in .ic-assets.json5.");
```

File: e2e/tests-dfx/assetscanister.bash [hunk 7]
```diff
@@ -1339,6 +1339,157 @@ EOF
...
+
+  # Security policy "hardened" defined for all assets
+  echo '[
+    {
+      "match": "**/*",
+      "security_policy": "hardened"
+    }
+  ]' > src/e2e_project_frontend/assets/.ic-assets.json5
+
+  assert_command dfx deploy
```
Comment (ericswanson-dfinity) [inline e2e/tests-dfx/assetscanister.bash:1467]: Related to suggestion below, I think this should
```suggestion
  assert_command_fail dfx deploy
```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Even though this isn't a breaking change, I think this PR should add migration notes to https://github.com/dfinity/sdk/tree/master/docs/migration for the next release, with a recommendation to replace all of the individually-defined headers in .ic-assets.json with use of a security policy.

File: src/canisters/frontend/ic-asset/src/sync.rs (hunk 5 above)
Comment (sesi200) [reply src/canisters/frontend/ic-asset/src/sync.rs:374]: I like that solution. Having hardened behave the same as standard except for the warning felt wrong

**Review by ericswanson-dfinity — APPROVED** [review]
